### PR TITLE
[SYCL][NFC] Employ the is_same_v helper in SYCL library

### PR DESCRIPTION
**Pull request by AGindinson**
Similarly to [intel/llvm@b469f03], refactor the usages of the
`is_same` type trait to reduce code verbosity. This particular
trait was chosen because of its popularity, and will probably
be the last in this refactoring series.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

### CODE REVIEW

**Comment by AGindinson** [comment]
`clang-format` seems to be firing a false alarm: at least 1 of the chunks is irrelevant to my changes. Other tool suggestions boil down to "`<template_parameter && template_parameter>` -> `<template_parameter&&template_parameter>`". Dear reviewers, I would appreciate some advice on how to proceed here.

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp [hunk 1]
```diff
@@ -209,7 +209,7 @@ template <typename U> constexpr bool is_type() { return false; }
 template <typename U, typename T, typename... Ts> constexpr bool is_type() {
   using UU = typename detail::remove_const_t<U>;
   using TT = typename detail::remove_const_t<T>;
-  return std::is_same<UU, TT>::value || is_type<UU, Ts...>();
+  return sycl::detail::is_same_v<UU, TT> || is_type<UU, Ts...>();
```
Comment (bader) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp:212]:
```suggestion
  return csd::is_same_v<UU, TT> || is_type<UU, Ts...>();
```

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp [hunk 2]
```diff
@@ -228,12 +228,13 @@ struct bitcast_helper {
 template <typename ToEltTy, typename FromEltTy, int FromN,
           typename = csd::enable_if_t<is_vectorizable<ToEltTy>::value>>
 ESIMD_INLINE typename detail::conditional_t<
-    std::is_same<FromEltTy, ToEltTy>::value, vector_type_t<FromEltTy, FromN>,
+    sycl::detail::is_same_v<FromEltTy, ToEltTy>,
```
Comment (bader) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp:231]:
```suggestion
    csd::is_same_v<FromEltTy, ToEltTy>,
```

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp [hunk 3]
```diff
@@ -228,12 +228,13 @@ struct bitcast_helper {
...
+    vector_type_t<FromEltTy, FromN>,
     vector_type_t<ToEltTy,
                   bitcast_helper<ToEltTy, FromEltTy, FromN>::nToElems()>>
 bitcast(vector_type_t<FromEltTy, FromN> Val) {
   // Noop.
-  if constexpr (std::is_same<FromEltTy, ToEltTy>::value)
+  if constexpr (sycl::detail::is_same_v<FromEltTy, ToEltTy>)
```
Comment (bader) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp:237]:
```suggestion
  if constexpr (csd::is_same_v<FromEltTy, ToEltTy>)
```

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp (hunk 1 above)
Comment (AGindinson) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_types.hpp]: Done, thanks for the suggestion. I've also applied the change to other `esimd/` files.

**Comment by AGindinson** [comment]
The MSVC build failure is somewhat resembling [this SO bug description](https://stackoverflow.com/questions/39579182/msvc-2013-type-is-not-a-member-of-stdenable-iffalse-void).

**Comment by bader** [comment]
> `clang-format` seems to be firing a false alarm: 1 of the chunks is irrelevant to my changes. Dear reviewers, I would appreciate some advice on how to proceed here.

What `clang-format` version do you use locally? I think we should try using new version in CI. I think we should `clang-format-9` today and it's quite old, so this issue might be already fixed in later versions.

**Comment by AGindinson** [comment]
> What `clang-format` version do you use locally? I think we should try using new version in CI. I think we should `clang-format-9` today and it's quite old, so this issue might be already fixed in later versions.

I'm basically using a self-built one, which is around version 12.

**Comment by bader** [comment]
> > What `clang-format` version do you use locally? I think we should try using new version in CI. I think we should `clang-format-9` today and it's quite old, so this issue might be already fixed in later versions.
> 
> I'm basically using a self-built one, which is around version 12. How would we go about updating the one used by CI?

I suggest using clang-format version used by CI system to avoid misalignments.

**Review by turinevgeny — DISMISSED** [review]
Looks good.

**Review by DenisBakhvalov — APPROVED** [review]
Changes in the ESIMD RT library look good.

**Comment by AGindinson** [comment]
Closing after the shift to C++14. Further refactoring should just use the `std::` variant of type traits.

**Pull request closed without merging**